kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraFactFlushManager.java

```diff
@@ -0,0 +1,80 @@
+/*
+ *
+ *  * Copyright 2023 Responsive Computing, Inc.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package dev.responsive.kafka.internal.db;
+
+import dev.responsive.kafka.internal.db.partitioning.TablePartitioner;
+import dev.responsive.kafka.internal.stores.RemoteWriteResult;
+import org.apache.kafka.common.utils.Bytes;
+
+public class CassandraFactFlushManager implements FlushManager<Bytes, Integer> {
+
+  private final String logPrefix;
+  private final CassandraFactTable table;
+  private final CassandraClient client;
+
+  private final TablePartitioner<Bytes, Integer> partitioner;
+  private final int kafkaPartition;
+
+  public CassandraFactFlushManager(
+      final CassandraFactTable table,
+      final CassandraClient client,
+      final int kafkaPartition
+  ) {
+    this.table = table;
+    this.client = client;
+    this.kafkaPartition = kafkaPartition;
+
+    partitioner = TablePartitioner.defaultPartitioner();
+    logPrefix = String.format("%s[%d] CassandraFactFlushManager", table.name(), kafkaPartition);
+  }
+
+  @Override
+  public TablePartitioner<Bytes, Integer> partitioner() {
+    return partitioner;
+  }
+
+  @Override
+  public RemoteWriter<Bytes, Integer> createWriter(final Integer tablePartition) {
+    return new FactSchemaWriter<>(
+        client,
+        table,
+        kafkaPartition,
+        tablePartition
+    );
+  }
+
+  @Override
+  public RemoteWriteResult<Integer> preFlush() {
+    return RemoteWriteResult.success(null);
+  }
+
+  @Override
+  public RemoteWriteResult<Integer> postFlush(final long consumedOffset) {
+    final int metadataPartition = partitioner.metadataTablePartition(kafkaPartition);
+    final var result = client.execute(table.setOffset(kafkaPartition, consumedOffset));
```

Comment:
just adding some inline comments to make sure this is safe (from first principles): if a batch write succeeds but we don't make it to `postFlush` the new writer will take over, recover from the old offsets, and rewrite them into the store during the restore phase.

let's say that this thread then wakes back up and successfully issues this offset flush - what happens? It'll overwrite the offset then fail b/c it was fenced by Kafka. Let's assume that for whatever reason the current writer (which has progressed past this point) then fails before it can overwrite this now zombie offset.

A new and third writer comes online and recovers from this offset, which may be a problem since it'll overwrite already processed data... 

I think this may be a bug in the current implementation as well (not introduced by this PR). We should just be safe and change the `setOffset` command to only set if the offset it is trying to write is greater than or equal to the one that's there.

Do you have bandwidth to address that as a follow up to this PR? (we should confirm my theory and that I'm not missing something, I thought that I had thought this situation through when designing fact tables but I may have missed it)

Reply:
> it'll overwrite already processed data...

Isn't this the basic premise of a fact table -- that this can happen but it's acceptable? eg for the deduplicator case, we may overwrite old values but we'll never "erase" data that already exists in the table. The assumption is that fact tables are only used in cases like that where the value can not necessarily be trusted to be up to date, but the "is this key present in the state store" answer is always consistent -- which it still would be if we accidentally restored some extra data?